pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java

```diff
@@ -330,48 +331,54 @@ protected void handleAuthChallenge(CommandAuthChallenge authChallenge) {
+            if (Arrays.equals(AuthData.REFRESH_AUTH_DATA_BYTES, authChallenge.getChallenge().getAuthData())) {
+                try {
+                    authenticationDataProvider = authentication.getAuthData(remoteHostName);
+                } catch (PulsarClientException e) {
+                    log.error("{} Error when refreshing authentication data provider: {}", ctx.channel(), e);
+                    connectionFuture.completeExceptionally(e);
+                    return;
+                }
             }
-        }
-
-        // mutual authn. If auth not complete, continue auth; if auth complete, complete connectionFuture.
-        try {
-            AuthData authData = authenticationDataProvider
-                .authenticate(AuthData.of(authChallenge.getChallenge().getAuthData()));
 
-            checkState(!authData.isComplete());
+            // mutual authn. If auth not complete, continue auth; if auth complete, complete connectionFuture.
+            try {
+                AuthData authData = authenticationDataProvider
+                        .authenticate(AuthData.of(authChallenge.getChallenge().getAuthData()));
 
-            ByteBuf request = Commands.newAuthResponse(authentication.getAuthMethodName(),
-                authData,
-                this.protocolVersion,
-                PulsarVersion.getVersion());
+                checkState(!authData.isComplete());
 
-            if (log.isDebugEnabled()) {
-                log.debug("{} Mutual auth {}", ctx.channel(), authentication.getAuthMethodName());
-            }
+                ByteBuf request = Commands.newAuthResponse(authentication.getAuthMethodName(),
+                        authData,
+                        protocolVersion,
+                        PulsarVersion.getVersion());
 
-            ctx.writeAndFlush(request).addListener(writeFuture -> {
-                if (!writeFuture.isSuccess()) {
-                    log.warn("{} Failed to send request for mutual auth to broker: {}", ctx.channel(),
-                        writeFuture.cause().getMessage());
-                    connectionFuture.completeExceptionally(writeFuture.cause());
+                if (log.isDebugEnabled()) {
+                    log.debug("{} Mutual auth {}", ctx.channel(), authentication.getAuthMethodName());
                 }
-            });
 
-            if (state == State.SentConnectFrame) {
-                state = State.Connecting;
+                ctx.writeAndFlush(request).addListener(writeFuture -> {
+                    if (!writeFuture.isSuccess()) {
+                        log.warn("{} Failed to send request for mutual auth to broker: {}", ctx.channel(),
+                                writeFuture.cause().getMessage());
+                        connectionFuture.completeExceptionally(writeFuture.cause());
+                    }
+                });
+
+                if (state == State.SentConnectFrame) {
+                    state = State.Connecting;
```

Comment:
Qs: is there any (added?) possibility of multiple concurrent calls, updating `state`, now that this will run asynchronously off the io thread?

Reply:
there is but it should be fine.  Currently,  the handler code can be concurrently executed when the number of pulsar-client-io threads is set to be more than 1